### ISSUE TITLE
generic collection types not assignable to variable of type GenericAlias

### ISSUE THREAD

**Issue by couteau** [microsoft/pyright#10527]
Pyright reports an error when attempting to assign a genericized builtin collection type to a variable declared to be of type `types.GenericAlias`.  In the below example, an error is reported when attempting to assign `list[int]` to a variable declared as `GenericAlias`, even though `GenericAlias` is literally defined as `type(list[int])`, and `isinstance(list[int], GenericAlias)` returns True. 

Relatedly, attempting to call a variable declared as `GenericAlias` reports that the object is not callable and that `GenericAlias.__call__` is unknown, even though all instances of `GenericAlias` are actually callable.

```python
from types import GenericAlias
from typing import reveal_type


def get_list_int_type() -> GenericAlias:
    print(type(list[int]))  # prints <class 'types.GenericAlias'>
    
    # Type type[list[int]] is not assignable to return type "GenericAlias"
    #    Type type[list[int]] is not assignable to type "GenericAlias" Pylance(reportReturnType)
    return list[int]


# Type type[list[int]] is not assignable to declared type "GenericAlias"
#    Type type[list[int]] is not assignable to type "GenericAlias" Pylance(reportAssignmentType)
t: GenericAlias = list[int]

get_list_int_type()

print(callable(t)) # prints 'True'
print(hasattr(t, '__call__')) # prints 'True'

lst = t() # Object of type GenericAlias is not callable
#    Attribute "__call__" is unknown Pylance(reportCallIssue)

reveal_type(lst) # Pyright reports Unknown, runtime type is list
```

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The type of the expression `list[int]` is `type[list[int]]`, and this type isn't assignable to `GenericAlias`. The class `GenericAlias` is an internal implementation detail used for some type expressions at runtime, so it shouldn't appear in type annotations.

For comparison, mypy generates the same error in this case.

The topic of runtime type form objects (like `GenericAlias` and `UnionType`) are discussed in [PEP 747](https://peps.python.org/pep-0747/) and the threads that informed this spec. PEP 747 is implemented in experimental form in pyright, and `TypeForm` was added to the type system for cases like this.

**Comment by jorenham** [comment]
> he type of the expression `list[int]` is `type[list[int]]`, and this type isn't assignable to `GenericAlias`.

Maybe it could be both? So by inferring `t = list[int]` as an instance of `type[list[int]] & GenericAlias` (the intersection type), or something along those lines.

**Comment by couteau** [comment]
I don't think it's accurate that `GenericAlias` is an internal implementation detail or that it is intended to be used only at runtime. It is a documented part of the standard library (see https://docs.python.org/3/library/stdtypes.html#types-genericalias), and per the documentation 
```
GenericAlias objects are intended primarily for use with type annotations.
```

The documentation also gives an example: "For example, list[int] is a GenericAlias object". And "GenericAlias objects are instances of the class `types.GenericAlias`".  

It seems strange that a type-checker would reject an assignment of an instance of a class to a variable annotated with that instance's class according to the python documentation.

**Comment by erictraut** [comment]
`TypeForm` is being added to the type system to support the scenario where type expressions are used as a value expression. The lengthly (multi-year!) discussions that informed PEP 747 highlighted the fact that this is a very complex topic, and there are many subtleties. It's not as easy as treating it as an intersection of two types.

I've already implemented experimental support for `TypeForm`. I don't plan to add alternative mechanisms for handling this. I think it's unlikely that other type checkers will either.
